apps/src/music/views/SoundsPanel2.tsx

```diff
@@ -0,0 +1,487 @@
+          const scrollPosition = distanceY - distanceY * easeOutSine(progress);
+
+          containerRef.current?.scroll(0, scrollPosition);
+
+          lastScrollPosition.current = scrollPosition;
+
+          scrollStep.current++;
+
+          if (scrollStep.current > delay + frames) {
+            // The automatic scroll has reached its destination.
+            scrollStep.current = undefined;
+          }
+        }
+      }
+    }, 1000 / 60);
+  }, [delay, distanceY, doScroll, frames]);
+
+  return (
+    <div id={id} className={className} ref={containerRefCallback}>
+      {children}
+    </div>
+  );
+};
+
+// SoundsPanel: a component used inside a Blockly dropdown that allows
+// for browsing folders and the sounds inside them.
+
+interface SoundsPanelProps {
+  library: MusicLibrary;
+  currentValue: string;
+  playingPreview: string;
+  showSoundFilters: boolean;
+  onSelect: (path: string) => void;
+  onPreview: (path: string) => void;
+}
+
+const SoundsPanel: React.FunctionComponent<SoundsPanelProps> = ({
+  library,
+  currentValue,
+  playingPreview,
+  showSoundFilters,
+  onSelect,
+  onPreview,
+}) => {
+  const folders = library.getAvailableSounds();
+  const libraryGroupPath = library.getPath();
+
+  const [selectedFolder, setSelectedFolder] = useState<SoundFolder>(
+    library.getAllowedFolderForSoundId(currentValue) || folders[0]
+  );
+  const [mode, setMode] = useState<Mode>('packs');
+  const [filter, setFilter] = useState<Filter>('all');
+
+  const currentFolderRef: React.MutableRefObject<HTMLDivElement | null> =
+    useRef(null);
+  const currentSoundRef: React.MutableRefObject<HTMLDivElement | null> =
+    useRef(null);
+
+  const isDefaultSoundSelected = currentValue === library.getDefaultSound();
+
+  const onModeChange = useCallback((value: Mode) => {
+    setMode(value);
+  }, []);
+
+  const onFilterChange = useCallback((value: Filter) => {
+    setFilter(value);
+  }, []);
+
+  // Initial render.
+  useEffect(() => {
+    // If the user has selected a non-default sound, then jump directly to it.
+    // If the default sound is selected, then the ScrollIntoView component will do
+    // an initial scroll instead.
+    if (!isDefaultSoundSelected) {
+      // This timeout allows the initial scroll-to-current-selection to work
+      // when wrapping the content with FocusLock.
+      setTimeout(() => {
+        currentFolderRef.current?.scrollIntoView();
+        currentSoundRef.current?.scrollIntoView();
+      }, 0);
+    }
+  }, [isDefaultSoundSelected]);
+
+  const currentFolderRefCallback = (ref: HTMLDivElement) => {
+    currentFolderRef.current = ref;
+  };
+
+  const currentSoundRefCallback = (ref: HTMLDivElement) => {
+    currentSoundRef.current = ref;
+  };
+
+  let possibleSoundEntries: SoundEntry[] = [];
+  let rightColumnSoundEntries: SoundEntry[] = [];
+
+  if (mode === 'packs') {
+    folders.sort((a, b) =>
+      a.restricted === b.restricted ? 0 : a.restricted ? -1 : 1
+    );
```

Comment:
just noting that this is sorting the array in place and confirming that we're okay doing that.

`Array.prototype.toSorted` is still too new for us to use. 😞

Reply:
Yes, this is intentional.

Reply:
Yes, this is intentional.